pennylane_sf/gbs.py

```diff
@@ -0,0 +1,246 @@
+# Copyright 2018-2020 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Strawberry Fields variational GBS device
+========================================
+
+**Module name:** :mod:`pennylane_sf.gbs`
+
+.. currentmodule:: pennylane_sf.gbs
+
+The Strawberry Fields variational GBS device provides a way to encode variational parameters into
+GBS so that the gradient with respect to the output probability distribution is accessible.
+
+Classes
+-------
+
+.. autosummary::
+   StrawberryFieldsGBS
+
+Code details
+~~~~~~~~~~~~
+"""
+from collections import OrderedDict
+
+import numpy as np
+import pennylane as qml
+import strawberryfields as sf
+from pennylane.operation import Probability
+from pennylane.wires import Wires
+from strawberryfields.ops import GraphEmbed, MeasureFock
+from strawberryfields.utils import all_fock_probs_pnr
+from thewalrus.quantum import find_scaling_adjacency_matrix as rescale
+from thewalrus.quantum import photon_number_mean_vector
+
+from .expectations import identity
+from .simulator import StrawberryFieldsSimulator
+
+
+class StrawberryFieldsGBS(StrawberryFieldsSimulator):
+    r"""StrawberryFields variational GBS device for PennyLane.
+
+    This device provides a method to embed variational parameters into GBS such that the analytic
+    gradient of the probability distribution is accessible.
+
+    Args:
+        wires (int): the number of modes to initialize the device in
+        analytic (bool): indicates if the device should calculate expectations
+            and variances analytically
+        cutoff_dim (int): Fock-space truncation dimension
+        backend (str): name of the remote backend to be used
+        shots (int): Number of circuit evaluations/random samples used
+            to estimate expectation values of observables. If ``analytic=True``,
+            this setting is ignored.
+        hbar (float): the convention chosen in the canonical commutation
+            relation :math:`[x, p] = i \hbar`
```

Comment:
Does `hbar` have any effect with this device? Since the adjacency matrix is in the `a^\dagger` basis, and the output is always Fock probabilities

Reply:
I'm not sure, but it may have an effect in how the covariance matrix is obtained from `A` and `n_mean`

Reply:
@trbromley it looks like all `StrawberryFieldsSimulator.probability()` does is set `sf.hbar = self.hbar`. We _could_ just set `self.hbar = 2` internally in the device as you suggest, but it sounds like best to keep it changeable by the user, as it could affect `n_mean`?

Reply:
I'm persuaded now - the choice of `hbar` is really only an internal parameter the user doesn't need to set it. I've removed it as an argument: https://github.com/PennyLaneAI/pennylane-sf/pull/47/files/1d4e04fa992f8c7782aac3ceb30e8773cc06073a..c9bc1fb805a615a3590ce099c4749c2fc91188ae